### ISSUE TITLE
"pip install Pillow" command not working. Returns an error requiring zlib

### ISSUE THREAD

**Issue by hero4barca** [python-pillow/Pillow#5868]
I ran "pip install Pillow " in a virtualenv but it returns an error about zlip dependency

I expected it to install Pillow in the virtual environment but it doesn't

 Pillow is a dependency for Django-filer which I obviously need for a djnago-cms project. It failed during installation as a dependency so I tried to do it myself - same error. Also I tried to download the precompiled wheel file ->(https://files.pythonhosted.org/packages/20/ec/15a263f2c65d71cf62aa767f774c2381077e07beb1e9309a94461ec1cd29/Pillow-8.4.0-cp39-cp39-win_amd64.whl) which should correspond to my environment but pip returns an error too:  "... not a  supported wheel on this platform"

P.S. This isn't the first time I have encountered this recently, I often try to file an alternative but have never been able to sort out this issue with pillow. The only other similar issue [here ](https://github.com/python-pillow/Pillow/issues/4158) suggest upgrading pip which doesn't apply to me as I am already using the latest version of PIP. Also, the current version of pillow is 8.4. and like I mentioned earlier, it is a dependency.

 Please, is it the case that Pillow doesn't play well with python3.9 or is there a way around this?? Thanks in advance

* OS: Windows 10
* Python: 3.9.6 64-bit
* Pillow: 8.4.0
* PIP: 21.3.1

###full error output#####
```
Running setup.py install for pillow ... error
    ERROR: Command errored out with exit status 1:
     command: 'c:\users\iffiok\documents\projects\django-cms\env\bin\python3.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Iffiok\\AppData\\Local\\Temp\\pip-install-g22ly81d\\pillow_0f7eb7d5f2b64b9ea643de25566269c3\\setup.py'"'"'; __file__='"'"'C:\\Users\\Iffiok\\AppData\\Local\\Temp\\pip-install-g22ly81d\\pillow_0f7eb7d5f2b64b9ea643de25566269c3\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Iffiok\AppData\Local\Temp\pip-record-f9lgxhke\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\users\iffiok\documents\projects\django-cms\env\include\site\python3.9\pillow'
         cwd: C:\Users\Iffiok\AppData\Local\Temp\pip-install-g22ly81d\pillow_0f7eb7d5f2b64b9ea643de25566269c3\
    Complete output (178 lines):
    running install
    running build
    running build_py
    creating build
    creating build\lib.mingw_x86_64-3.9
    creating build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\BdfFontFile.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\BlpImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\BmpImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\BufrStubImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ContainerIO.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\CurImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\DcxImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\DdsImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\EpsImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ExifTags.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\features.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\FitsStubImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\FliImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\FontFile.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\FpxImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\FtexImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\GbrImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\GdImageFile.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\GifImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\GimpGradientFile.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\GimpPaletteFile.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\GribStubImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\Hdf5StubImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\IcnsImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\IcoImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\Image.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageChops.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageCms.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageColor.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageDraw.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageDraw2.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageEnhance.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageFile.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageFilter.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageFont.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageGrab.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageMath.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageMode.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageMorph.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageOps.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImagePalette.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImagePath.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageQt.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageSequence.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageShow.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageStat.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageTk.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageTransform.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImageWin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\ImtImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\IptcImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\Jpeg2KImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\JpegImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\JpegPresets.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\McIdasImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\MicImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\MpegImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\MpoImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\MspImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PaletteFile.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PalmImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PcdImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PcfFontFile.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PcxImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PdfImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PdfParser.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PixarImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PngImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PpmImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PsdImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PSDraw.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\PyAccess.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\SgiImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\SpiderImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\SunImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\TarIO.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\TgaImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\TiffImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\TiffTags.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\WalImageFile.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\WebPImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\WmfImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\XbmImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\XpmImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\XVThumbImagePlugin.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\_binary.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\_tkinter_finder.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\_util.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\_version.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\__init__.py -> build\lib.mingw_x86_64-3.9\PIL
    copying src\PIL\__main__.py -> build\lib.mingw_x86_64-3.9\PIL
    running egg_info
    writing src\Pillow.egg-info\PKG-INFO
    writing dependency_links to src\Pillow.egg-info\dependency_links.txt
    writing top-level names to src\Pillow.egg-info\top_level.txt
    adding license file 'LICENSE' (matched pattern 'LICEN[CS]E*')
    reading manifest file 'src\Pillow.egg-info\SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    warning: no files found matching '*.c'
    warning: no files found matching '*.h'
    warning: no files found matching '*.sh'
    warning: no previously-included files found matching '.appveyor.yml'
    warning: no previously-included files found matching '.clang-format'
    warning: no previously-included files found matching '.coveragerc'
    warning: no previously-included files found matching '.editorconfig'
    warning: no previously-included files found matching '.readthedocs.yml'
    warning: no previously-included files found matching 'codecov.yml'
    warning: no previously-included files matching '.git*' found anywhere in distribution
    warning: no previously-included files matching '*.pyc' found anywhere in distribution
    warning: no previously-included files matching '*.so' found anywhere in distribution
    no previously-included directories found matching '.ci'
    writing manifest file 'src\Pillow.egg-info\SOURCES.txt'
    running build_ext
    dllwrap: WARNING: dllwrap is deprecated, use gcc -shared or ld -shared instead



    The headers or library files could not be found for zlib,
    a required dependency when compiling Pillow from source.

    Please see the install instructions at:
       https://pillow.readthedocs.io/en/latest/installation.html

    Traceback (most recent call last):
      File "C:\Users\Iffiok\AppData\Local\Temp\pip-install-g22ly81d\pillow_0f7eb7d5f2b64b9ea643de25566269c3\setup.py", line 978, in <module>
        setup(
      File "c:\users\iffiok\documents\projects\django-cms\env\lib\python3.9\site-packages\setuptools\__init__.py", line 153, in setup
        return distutils.core.setup(**attrs)
      File "C:\msys64\mingw64\lib\python3.9\distutils\core.py", line 148, in setup
        dist.run_commands()
      File "C:\msys64\mingw64\lib\python3.9\distutils\dist.py", line 966, in run_commands
        self.run_command(cmd)
      File "C:\msys64\mingw64\lib\python3.9\distutils\dist.py", line 985, in run_command
        cmd_obj.run()
      File "c:\users\iffiok\documents\projects\django-cms\env\lib\python3.9\site-packages\setuptools\command\install.py", line 61, in run
        return orig.install.run(self)
      File "C:\msys64\mingw64\lib\python3.9\distutils\command\install.py", line 547, in run
        self.run_command('build')
      File "C:\msys64\mingw64\lib\python3.9\distutils\cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "C:\msys64\mingw64\lib\python3.9\distutils\dist.py", line 985, in run_command
        cmd_obj.run()
      File "C:\msys64\mingw64\lib\python3.9\distutils\command\build.py", line 135, in run
        self.run_command(cmd_name)
      File "C:\msys64\mingw64\lib\python3.9\distutils\cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "C:\msys64\mingw64\lib\python3.9\distutils\dist.py", line 985, in run_command
        cmd_obj.run()
      File "c:\users\iffiok\documents\projects\django-cms\env\lib\python3.9\site-packages\setuptools\command\build_ext.py", line 79, in run
        _build_ext.run(self)
      File "C:\msys64\mingw64\lib\python3.9\distutils\command\build_ext.py", line 341, in run
        self.build_extensions()
      File "C:\Users\Iffiok\AppData\Local\Temp\pip-install-g22ly81d\pillow_0f7eb7d5f2b64b9ea643de25566269c3\setup.py", line 790, in build_extensions    
        raise RequiredDependencyException(f)
    __main__.RequiredDependencyException: zlib

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\Iffiok\AppData\Local\Temp\pip-install-g22ly81d\pillow_0f7eb7d5f2b64b9ea643de25566269c3\setup.py", line 1037, in <module>
        raise RequiredDependencyException(msg)
    __main__.RequiredDependencyException:

    The headers or library files could not be found for zlib,
    a required dependency when compiling Pillow from source.

       https://pillow.readthedocs.io/en/latest/installation.html


    ----------------------------------------
ERROR: Command errored out with exit status 1: 'c:\users\iffiok\documents\projects\django-cms\env\bin\python3.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Iffiok\\AppData\\Local\\Temp\\pip-install-g22ly81d\\pillow_0f7eb7d5f2b64b9ea643de25566269c3\\setup.py'"'"'; __file__='"'"'C:\\Users\\Iffiok\\AppData\\Local\\Temp\\pip-install-g22ly81d\\pillow_0f7eb7d5f2b64b9ea643de25566269c3\\setup.py'"'"';f = getattr(tokenize, 
'"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Iffiok\AppData\Local\Temp\pip-record-f9lgxhke\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\users\iffiok\documents\projects\django-cms\env\include\site\python3.9\pillow' Check the logs for full command output.
```

**Comment by nulano** [comment]
> ```
>File "C:\msys64\mingw64\lib\python3.9\distutils\core.py", line 148, in setup
>    dist.run_commands()
>```

It seems you are running on MSYS2/MinGW64.

To install Pillow from source you can follow the instructions here: https://pillow.readthedocs.io/en/stable/installation.html#building-on-windows-using-msys2-mingw
Your specific error would likely be fixed by running `pacman -S mingw-w64-x86_64-zlib`, but read the link to install all optional dependencies also if you wish.

You can also install Pillow for the system from the MSYS2 package repository (maintained by them) with `pacman -S mingw-w64-python-pillow`. As of writing this that should install [version `8.4.0`](https://packages.msys2.org/base/mingw-w64-python-pillow), but the package was quite outdated not that long ago.

**Comment by nulano** [comment]
> Please, is it the case that Pillow doesn't play well with python3.9 or is there a way around this?? Thanks in advance

This is because you are using MSYS2/MinGW64 Python. Pillow doesn't release precompiled binary packages for this platform. See my previous comment to either install from source or the MSYS2 package repository.

**Comment by hero4barca** [comment]
Hi,
 Thanks for your prompt response. Per installing zlip on MinG4, I did follow this instruction previously and confirm that I do have zlip installed. See screenshot below

`
![image](https://user-images.githubusercontent.com/15052725/144748614-7fe11dbb-b253-467f-a2c2-eec1dd70f3b5.png)

It one of the things that really confuse me about this. Also if I open a normal cmd, I do have pillow installed, see scree shot below:

![image](https://user-images.githubusercontent.com/15052725/144749097-db3c041d-2ae8-4b5b-a411-49384dd813a1.png)
As seen in the screenshot above, I can install and uninstall in the screenshot above easily. Enters the virtualenv, once I create a venv and try to install pillow there, then  zlib error  pops up....

**Comment by nulano** [comment]
> Also if I open a normal cmd, I do have pillow installed, see scree shot below:
>
> As seen in the screenshot above, I can install and uninstall in the screenshot above easily. Enters the virtualenv, once I create a venv and try to install pillow there, then zlib error pops up....

You have Pillow installed in a native Windows version of Python 3.7, but your venv is based on Python 3.9 from MSYS2/mingw64.

> Per installing zlip on MinG4, I did follow this instruction previously and confirm that I do have zlip installed. See screenshot below

Hmm, I would expect that to work.
Could you try activating your venv in a mingw64 terminal window and trying to install Pillow there?

**Comment by hero4barca** [comment]
@nulano
Thanks for bringing that up!!! I was already taking a look at difference, putting it like that, you were spot on. Will just put it down here in the unlikely event that it's useful to anyone else.

Despite never really intending to use  MinGW for python development, it is on my path for some c++ compilations that needs it. Now, since it comes with python, each time I created a virtualenv (since setting up MinGW and adding it to path), it used the python in MinGw as opposed to my default (preferred ) python installation. Creating the virtualenv with the path to my original python installation gave me a way AROUND this: `<path/to/python> -m venv env_name. 
`
I guess removing MinGW from path could also work.  

It's a way around because I haven't figured out how to fix the zlib issue especially since I did have the library installed on Pacman while I was getting the error. However, I can proceed without having to fix it because it isn't a problem unless you are using the python from minGW.
